### ISSUE TITLE
How to properly reimplement base hooks?

### ISSUE THREAD

**Issue by dramalho** [dblock/slack-ruby-bot#50]
Say I want to perform an action once I have confirmation that the client is connected. 

Looking at [the api documentation](https://api.slack.com/events/hello) it seems like the `hello` event is exactly what I want, so I implement a simple hook for that event and I include it in my class (in my case a subclass of App).

When I run the code though - a simple logger line to check that it's being called properly - I get two calls instead of one. The reason is, [Hooks::Base](https://github.com/dblock/slack-ruby-bot/blob/master/lib/slack-ruby-bot/hooks/base.rb) pushes the included class name into an array, and `Server` already includes a base `Hello` hook,  so the hook stack looks like

````ruby
[
    [0] :hello,
    [1] :message,
    [2] :hello
]
````

However, because these are just methods on the App object (and each new `include` will overwrite these) we basically end up with the same hook method being called twice.

So, it might have been the way I did it (extending `App`, including my own hook) and that might not be the right approach - I couldn't find another though. Or, it could be that the `Hook::Base` simply dedups the hook stack and this gets sorted . 

Not sure, though I run this by you first :)

**Comment by dramalho** [comment]
For now I'm doing this in my initializer

```ruby
    def initialize(options = {})
      self.class.hooks.uniq!  # Dedup hooks

      super
    end
````

but it's mostly hacky :)

**Comment by dblock** [comment]
I think the way I implemented this is really not very smart ;) If you look at https://github.com/dblock/slack-ruby-bot/blob/7f54f00cbcd17ff4fba0a3c95de332f9e81ec4cd/lib/slack-ruby-bot/server.rb#L96 we assume a lot of things here. First that the class matches the method, which makes it impossible to include multiple things like the example you give. I say we fix it properly.

I would either kill SlackRubyBot::Hooks::Base altogether or just deprecate that way of doing things in favor of something much more elegant:

```ruby
class MyBotServer < Server
  on :hello do
     # evaluates in the context of this class
  end
end
```

I did a similar refactor in the client [here](https://github.com/dblock/slack-ruby-client/commit/d612c35ee7b8f59a68a35f40a20fc3c296c77cd1) around events and it turned out to be a lot cleaner.

Want to give it a shot?

**Comment by dramalho** [comment]
So, to be honest, I like both syntactic sugar approach (for small enough bots) and the one-class per hook implementation (like the commands) for more complex stuff, it helps keep the code decent sized so, again, like the commands, I would rather have both options. Also, I don't see a particular reason why two different handlers couldn't be notified of the same event, so I would put that in. 

I also understand that at least for the `message` hook, it's great to have a default implementation there out of the box, but it probably shouldn't stop some happy person from rolling out it's own - without re-implementing ::Server - so maybe add a way to unregister / clean hooks.

re: magic / assumptions, it can be made a bit sane I guess, descendants of Base, class name matches hook name, we can go either way, magic (class exists) or explicit (someone registers a hook class) and then each class can have a common interface (i.e. hook classes receive `call` with a bunch of params)

Sorry, I'm just rambling a bit on the train :) , but I would roughly enjoy having:

* Multiple handlers per hook
* Explicit way to remove handlers (say the default ones)
* Syntactic command-like sugar for simpler single file bots, but also individual handler classes
* Still be able to just extend `Server` / `App` and tailor configurations

And yeah, I don't mind taking a stab at this, but it's your baby, I want to make sure I move in a direction you agree with :) , so what do you think?

**Comment by dblock** [comment]
Everything you say makes total sense. Would love to see you contribute the missing parts!

### Multiple handlers per hook

Definitely. A hook is just a callback mechanism, so being able to do multiple things in it would be great. 

### Explicit way to remove handlers (say the default ones)

I think having the default add them and client code remove them is hacky, I would rather add a way not to add the default ones via some `configure` or similar call.

### Syntactic command-like sugar for simpler single file bots, but also individual handler classes

Yes.

### Still be able to just extend Server / App and tailor configurations

:+1:

**Comment by dramalho** [comment]
| I think having the default add them and client code remove them is hacky, I would rather add a way not to add the default ones via some configure or similar call.

yeah, I'm just thinking of the scenario where [this](https://github.com/dblock/slack-ruby-bot/blob/master/lib/slack-ruby-bot/hooks/message.rb) is there from the start and you can actually get a basic bot working out of the box. 

I'll have a think about that one in particular, agree that undoing something super did feels hacky

**Comment by dramalho** [comment]
I'll carve some time to work on this, if I can I'll break it into separate PR's but I need to structure it first :) 

like I said, happy to help :)

**Comment by dramalho** [comment]
actually, happy to give something back in return for the work that you did :)

**Comment by dblock** [comment]
:+1:

**Comment by dramalho** [comment]
Naive implementation #1 https://github.com/dblock/slack-ruby-bot/compare/master...dramalho:hook-improvements, on my bot I can then configure handlers like

````ruby
{
        hello: [
          SlackRubyBot::Hooks::Hello.new(logger),
          Bot::Events::Hello.new
        ],
        message: SlackRubyBot::Hooks::Message.new(logger),
        user_presence: Bot::Events::UserChange.new,
        presence_change: Bot::Events::PresenceChange.new
      }
````

mehish

**Comment by dblock** [comment]
Other than some naming suggestions I think that's actually an improvement, I would turn this into a PR.

**Comment by dramalho** [comment]
I'll throw specs at it. Naming suggestions welcome, it's the bane of my
existence

**Comment by dblock** [comment]
I only have caching suggestions :)

Generally I think you should keep names short. For example, `register_handler` would belong in a class called `Registry::Handlers` instead of just `Registry`, and then the method becomes `register` or just `add`. That's just to illustrate, not sure whether that's right here.

**Comment by dramalho** [comment]
Yes, no one will ever know David for it's ability to write short names :D . `register` sounds cool, the context is narrow anyway, so people won't get confused by what they're registering. 

Perfect, I'll work on adding specs and clean this up .

There's one annoying factor though, people who upgrade to this will potentially have their hook handlers break on them. Not sure if I can make it fully backwards compatible (I can try)

**Comment by dblock** [comment]
I would try, but not more than that, for backwards compatibility. Few people do custom hooks, and we have a clear https://github.com/dblock/slack-ruby-bot/blob/master/UPGRADING.md document that should explain what's going on. This stuff is pretty new, we can break things for now.

**Comment by dramalho** [comment]
Quick update, just so you don't think I disappeared :) 

I obviously broke every spec in the system, so I [changed the rspec matcher a bit](https://github.com/dramalho/slack-ruby-bot/commit/5602bb0eac2c56c837c1e4600470c120e79a37d2) to use the Message hook directly - although I think there's a bit too much rolling on that `message` Hook, or at least I would like specs to me more explicit about it, I won't do much more refactoring now  . 
I need to tackle the block like commands too, but given they are Procs and I can `call` them, it should just be a matter of simply adding them to the registry, but I haven't done that yet.

And then my own specs :)

**Comment by dblock** [comment]
:+1:

**Comment by dblock** [comment]
It's looking good. Something still bugs me about the name `Registry`, it's very Windows. What do you think of just `Set`?

**Comment by dramalho** [comment]
`Hooks::Set` is fine and dandy :P , I'll change that up

**Comment by dramalho** [comment]
and yeah, it's been yeeeeeeeeeeeeears since I last used Windows, but I get what you mean, I get _exactly_ what you mean :P

**Comment by dblock** [comment]
![](https://media.giphy.com/media/AliO819Kp3aus/giphy.gif)

**Comment by dramalho** [comment]
OMG now I want to write a ICQ bot to run on my windows 95 box :D

**Comment by dramalho** [comment]
Cool, specs are passing and all that jazz, I'll open a PR and we can more the discussion there :)

**Comment by dramalho** [comment]
Closed by the power of [#54](https://github.com/dblock/slack-ruby-bot/pull/54)
